impeller/compiler/spirv_compiler.cc

```diff
@@ -0,0 +1,305 @@
+    COMPILER_ERROR(stream) << "Invalid sources for SPIRV Compiler.";
+    return nullptr;
+  }
+
+  shaderc::Compiler spv_compiler;
+  if (!spv_compiler.IsValid()) {
+    COMPILER_ERROR(stream) << "Could not initialize the "
+                           << SourceLanguageToString(options_.source_language)
+                           << " to SPIRV compiler.";
+    return nullptr;
+  }
+
+  const auto shader_kind = ToShaderCShaderKind(options_.type);
+
+  if (shader_kind == shaderc_shader_kind::shaderc_glsl_infer_from_source) {
+    COMPILER_ERROR(stream) << "Could not figure out shader stage.";
+    return nullptr;
+  }
+
+  auto result = std::make_shared<shaderc::SpvCompilationResult>(
+      spv_compiler.CompileGlslToSpv(
+          reinterpret_cast<const char*>(sources_->GetMapping()),  // source_text
+          sources_->GetSize(),                // source_text_size
+          shader_kind,                        // shader_kind
+          options_.file_name.c_str(),         // input_file_name
+          options_.entry_point_name.c_str(),  // entry_point_name
+          spirv_options                       // options
+          ));
+  if (result->GetCompilationStatus() !=
+      shaderc_compilation_status::shaderc_compilation_status_success) {
+    COMPILER_ERROR(stream) << SourceLanguageToString(options_.source_language)
+                           << " to SPIRV failed; "
+                           << ShaderCErrorToString(
+                                  result->GetCompilationStatus())
+                           << ". " << result->GetNumErrors() << " error(s) and "
+                           << result->GetNumWarnings() << " warning(s).";
+    // It should normally be enough to check that there are errors or warnings,
+    // but some cases result in no errors or warnings and still have an error
+    // message. If there's a message we should print it.
+    if (result->GetNumErrors() > 0 || result->GetNumWarnings() > 0 ||
+        !result->GetErrorMessage().empty()) {
+      COMPILER_ERROR_NO_PREFIX(stream) << result->GetErrorMessage();
+    }
+    return nullptr;
+  }
+
+  if (!result) {
+    COMPILER_ERROR(stream) << "Could not fetch SPIRV from compile job.";
+    return nullptr;
+  }
+
+  const auto data_length = (result->cend() - result->cbegin()) *
+                           sizeof(decltype(result)::element_type::element_type);
+
+  return std::make_unique<fml::NonOwnedMapping>(
+      reinterpret_cast<const uint8_t*>(result->cbegin()),  //
+      data_length,                                         //
+      [result](auto, auto) {}                              //
+  );
+}
+
+std::string SPIRVCompiler::GetSourcePrefix() const {
+  std::stringstream stream;
+  stream << options_.file_name << ": ";
+  return stream.str();
+}
+
+static void SetDefaultLimitations(shaderc::CompileOptions& compiler_opts) {
+  using Limit = std::pair<shaderc_limit, int>;
+  static constexpr std::array<Limit, 83> limits = {
```

Comment:
Not sure why only clang on Windows doesn't like this line. Maybe it wants some explicit `#include`?

Reply:
Yeah, I moved this to another TU but didn't move the `#include<array>`.